src/keri/core/kraming.py

```diff
@@ -0,0 +1,231 @@
+    """TimelinessCache is responsible for preventing replay attacks in KERI/KRAM by ensuring
+    messages are timely and processed in a strictly monotonically ordered fashion.
+
+    It maintains:
+    1. A Lagging Window Size Table - to determine validity windows for different message types
+    2. A Replay Cache Table - to store timestamps of previously validated messages
+    """
+
+    def __init__(self, db, defaultWindowSize=3.0, defaultDriftSkew=1.0):
+        """Initialize the TimelinessCache.
+
+        Parameters:
+            db: Database instance that contains the IoSetSuber at db.time
+            defaultWindowSize (float): Default window size in seconds
+            defaultDriftSkew (float): Default drift skew in seconds
+        """
+        self.defaultWindowSize = defaultWindowSize
+        self.defaultDriftSkew = defaultDriftSkew
+
+        self.db = db
+
+    def setWindowParameters(self, aid, windowSize=None, driftSkew=None, messageType=None):
+        """Set window parameters for given autonomic identifier and message type.
+        """
+        # TODO: Implement message type as part of the window parameters
+        windowSize = windowSize or self.defaultWindowSize
+        driftSkew = driftSkew or self.defaultDriftSkew
+
+        # Serialize the tuple as JSON bytes for storage
+        windowTuple = (windowSize, driftSkew)
+        serialized = json.dumps(windowTuple).encode('utf-8')
+
+        self.db.kram.pin(aid, [serialized])
+
+
+    def getWindowParameters(self, aid, messageType=None):
+        """Get window parameters for given autonomic identifier and message type.
+        Falls back to default values if no entry exists for the aid.
+
+        Parameters:
+            aid (str): autonomic identifier
+            messageType (str | None): message type identifier. None for now, but will
+                be used to determine the window size for a given message type
+
+        Returns:
+            tuple: (windowSize, driftSkew) as floats in seconds
+        """
+        # TODO: Implement message type as part of the window parameters
+        try:
+            # Try to get the stored parameters
+            storedData = self.db.kram.getLast(aid)
+            if storedData is not None:
+                # Deserialize from JSON bytes
+                windowTuple = json.loads(storedData.decode('utf-8'))
+                return windowTuple[0], windowTuple[1]
+        except Exception:
+            pass
+
+        # Fallback to defaults
+        return self.defaultWindowSize, self.defaultDriftSkew
+
+
+    def _constructCacheKey(self, serder):
+        """Construct the key for the Replay Cache Table.
+
+        Parameters:
+            serder: The SerderKERI instance containing the message
+
+        Returns:
+            str: The key for the Replay Cache Table
+        """
+        # TODO: Implement message type as part of the key (serder.ilk)
+        sad = serder.sad
+        sourceAid = sad.get("i", "")
```

Comment:
If you do `ilk = sad.get("t", "")` and then return `(sourceAid, ilk)` then you can get support the message type.